test/unit/modules/blocks/chain.js

```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
+								block,
+								{ id: 1, type: 1 },
+								'a1',
+								() => {
+									done();
+								}
+							);
+						});
+					});
+				});
+			});
 		});
 	});
 
-	describe('deleteAfterBlock', () => {
-		it('should call library.db.query');
-
-		it('should call library.db.query with sql.deleteAfterBlock');
-
-		it('should call library.db.query with {id: blockId}');
-
-		describe('when library.db.query fails', () => {
-			it('should call logger.error');
-
-			it('should call logger.error with err.stack');
-
-			it('should call callback with "Blocks#deleteAfterBlock error"');
-		});
+	describe('__private.undoUnconfirmedListStep', () => {
+		describe('modules.transactions.undoUnconfirmedList', () => {
+			describe('when fails', () => {
+				beforeEach(() => {
+					return modules.transactions.undoUnconfirmedList.callsArgWith(
+						0,
+						'undoUnconfirmedList-ERR',
+						null
+					);
+				});
 
-		describe('when library.db.query succeeds', () => {
-			it('should call callback with error = null');
+				afterEach(() => {
+					expect(loggerStub.error.args[0][0]).to.be.equal(
+						'Failed to undo unconfirmed list'
+					);
+					return expect(loggerStub.error.args[0][1]).to.be.equal(
+						'undoUnconfirmedList-ERR'
+					);
+				});
+				it('should call a callback with error', done => {
+					__private.undoUnconfirmedListStep(err => {
+						expect(err).to.equal('Failed to undo unconfirmed list');
+						done();
+					});
+				});
+			});
 
-			it('should call callback with result = res');
+			describe('when succeeds', () => {
+				beforeEach(() => {
+					return modules.transactions.undoUnconfirmedList.callsArgWith(
+						0,
+						null,
+						true
+					);
+				});
+				it('should call a callback with no error', done => {
+					__private.undoUnconfirmedListStep(err => {
+						expect(err).to.be.undefined;
+						done();
+					});
+				});
+			});
 		});
 	});
 
-	describe('applyGenesisBlock', () => {
-		it('should sort transactions after type');
-
-		it('should call modules.blocks.utils.getBlockProgressLogger');
-
-		it(
-			'should call modules.blocks.utils.getBlockProgressLogger with block.transactions.length'
-		);
+	describe('__private.applyUnconfirmedStep', () => {
+		describe('when block.transactions is undefined', () => {
+			it('should return rejected promise with error', done => {
+				__private
+					.applyUnconfirmedStep(
+						{ id: 6, height: 6, transactions: undefined },
+						dbStub.tx
+					)
+					.catch(err => {
+						expect(err.message).to.equal(
+							'expecting an array or an iterable object but got [object Null]'
+						);
+						done();
+					});
+			});
+		});
 
-		it(
-			'should call modules.blocks.utils.getBlockProgressLogger with block.transactions.length / 100'
-		);
+		describe('when block.transactions is empty', () => {
+			it('should return resolved promise with no error', done => {
+				__private
+					.applyUnconfirmedStep(
+						{ id: 6, height: 6, transactions: [] },
```

Comment:
we have 3-4 type of blocks. We should define these blocks and reuse them in these tests.